axum/src/response/sse.rs

```diff
@@ -407,6 +406,48 @@ impl Event {
     }
 }
 
+impl std::io::Write for EventDataWrite<'_> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        if let Some(flags) = self.flags.take() {
+            if flags.contains(EventFlags::HAS_DATA) {
+                panic!("Called `Event::data*` multiple times");
+            }
+
+            self.buffer.write_all(b"data: ")?;
+
+            flags.insert(EventFlags::HAS_DATA);
+        }
+
+        let mut last_split = 0;
+        for delimiter in memchr::memchr2_iter(b'\n', b'\r', buf) {
+            self.buffer.write_all(&buf[last_split..=delimiter])?;
+            self.buffer.write_all(b"data: ")?;
+            last_split = delimiter + 1;
+        }
+        self.buffer.write_all(&buf[last_split..])?;
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.buffer.flush()
+    }
+}
+
+impl Drop for EventDataWrite<'_> {
+    fn drop(&mut self) {
+        // ASSUMPTION: data has been written if flags is None
+        // (see impl of `std::io::Write` for `EventDataWrite`)
+        if self.flags.is_none() {
```

Comment:
The trouble with this is that we can’t technically rely on `Drop` being called. But on the other hand, getting a corrupted SSE stream from `forget`ing this type, which no one should be doing anyway, does not seem like the worst outcome. So maybe just add a comment here explaining the tradeoff?